[AArch64][GlobalISel] Improve MULL generation

davemgreen opened the pull request:
This splits the existing post-legalize lowering of vector umull/smull into two parts - one to perform the optimization of mul(ext,ext) -> mull and one to perform the v2i64 mul scalarization. The mull part is moved to post legalizer combine and has been taught a few extra tricks from SDAG, using known bits to convert mul(sext, zext) or mul(zext, zero-upper-bits) into umull. This can be important to prevent v2i64 scalarization of muls.

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp:
```diff
@@ -438,6 +438,109 @@ void applyCombineMulCMLT(MachineInstr &MI, MachineRegisterInfo &MRI,
+  bool IsSExt1 =
+      I1Opc == TargetOpcode::G_SEXT || I1Opc == TargetOpcode::G_ANYEXT;
+  bool IsSExt2 =
+      I2Opc == TargetOpcode::G_SEXT || I2Opc == TargetOpcode::G_ANYEXT;
```
I think you could write a lambda function taking TargetOpcode and return bool to check the condition. You can avoid repetition of `TargetOpcode::G_SEXT || TargetOpcode::G_ANYEXT`

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp:
```diff
@@ -438,6 +438,109 @@ void applyCombineMulCMLT(MachineInstr &MI, MachineRegisterInfo &MRI,
+  bool IsZExt1 =
+      I1Opc == TargetOpcode::G_ZEXT || I1Opc == TargetOpcode::G_ANYEXT;
+  bool IsZExt2 =
+      I2Opc == TargetOpcode::G_ZEXT || I2Opc == TargetOpcode::G_ANYEXT;
```
I think you could write a lambda function taking TargetOpcode and return bool to check the condition. You can avoid repetition of `TargetOpcode::G_ZEXT || TargetOpcode::G_ANYEXT`

madhur13490 commented on llvm/lib/Target/AArch64/AArch64Combine.td:
```diff
@@ -208,11 +208,19 @@ def mul_const : GICombineRule<
+def mull_matchdata : GIDefMatchData<"std::tuple<bool, Register, Register>">;
+def extmultomull : GICombineRule<
+  (defs root:$root, mull_matchdata:$matchinfo),
+  (match (wip_match_opcode G_MUL):$root,
```
Can we please avoid using `wip_match_opcode`. I tried to get rid of it from this file for most of the things so introducing more would be a step backwards :)

tschuett commented on llvm/lib/Target/AArch64/AArch64Combine.td:
```diff
@@ -208,11 +208,19 @@ def mul_const : GICombineRule<
+def extmultomull : GICombineRule<
+  (defs root:$root, mull_matchdata:$matchinfo),
+  (match (G_MUL $dst, $src1, $src2):$root,
+          [{ return matchExtMulToMULL(*${root}, MRI, KB, ${matchinfo}); }]),
```
Could you take a look at push_sub_through_zext, which is at the top of this file?

davemgreen commented:
I've updated this to be more like the SDAG equivalent in selectUmullSmull, where it looks at the known buts / sign bits for converting to umull/smull, especially for 64bit types.

davemgreen reviewed (commented):
Rebase and ping - thanks.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp:
```diff
@@ -438,6 +438,123 @@ void applyCombineMulCMLT(MachineInstr &MI, MachineRegisterInfo &MRI,
+  if (Src2Ty.getScalarSizeInBits() * 2 != DstTy.getScalarSizeInBits())
+    Src2Reg = B.buildExtOrTrunc(ExtOpc, {HalfDstTy}, {Src2Reg}).getReg(0);
+
+  B.setInstrAndDebugLoc(MI);
```
I think this is pre-set before any apply action now

davemgreen committed:
[AArch64][GlobalISel] Improve MULL generation

This splits the existing post-legalize lowering of vector umull/smull into two
parts - one to perform the optimization of mul(ext,ext) -> mull and one to
perform the v2i64 mul scalarization. The mull part is moved to post legalizer
combine and has been taught a few extra tricks from SDAG, using known bits to
convert mul(sext, zext) or mul(zext, zero-upper-bits) into umull. This can be
important to prevent v2i64 scalarization of muls.

davemgreen commented:
I've rebased over the changes in GISelValueTracking. Thanks

kazutakahirata commented:
@davemgreen I've landed e65faed187485db3638aa6415f7e9df053623dcf to fix a warning from this PR.  Thanks!
